Add Package Control docs (#134)
sublimetext-io/docs.sublimetext.io: docs/reference/pc_repository.md

---
title: "Package Control: repository.json"
---

<!-- Original: -->
<!-- https://github.com/wbond/package_control/blob/master/example-repository.json -->

# Examples for the repository.json

The default repository.json is managed in the package_control_channel repo:

> https://github.com/wbond/package_control_channel/blob/master/repository.json

Author: `package_control_channel` isn't "code" per se, it's the name of a project. Maybe we should call it "Package Control Default Channel" instead of calling it by the repo name.

					"platforms": ["windows-x64", "linux-x64", "osx-x64"],
					"sublime_text": ">=4107",
					"tags": "st4107-"
				}
			]
		},
		{
			"name": "My Package",
			"description": "A locally hosted Sublime Text package",
			"author": "John Doe",
			"issues": "https://company.com/software/issues",
			"releases": [
				{
					"platforms": ["windows-x64"],
					"version": "1.0.0",
					"url": "file:///c:/absolute/path/to/My Package.sublime-package"
				},
				{
					"platforms": ["linux-x64"],
					"version": "1.0.0",
					"url": "./path/relative/to/repository_json/My Package.sublime-package"
				}
			]
		}
	]
}
```

## `releases`

All packages must have one or more `releases`.
Releases reference tags named in accordance with [semver](https://semver.org).
The value for `tags` can be simply `true` or a string prefix.
The latter is specifically useful when targeting older versions of ST
with a specific range of tags.

Note that previously releases based on branches instead of tags was supported,
but this has been deprecated.

### `sublime_text`

A release MUST contain a "sublime_text" version selector.

> Use `*` for all versions,

Reviewer: ```suggestion
Use "\*" for all versions,
```

Or change all the values to code spans.

> or "<3000", "3500-4000", etc.

Reviewer: ```suggestion
or "<3000", "3500 - 4000", etc.
```

I don't know if the spaces are required, but we use them everywhere else there's a range.

It allows specifying a single platform, or a list of platforms.
Valid platform identifiers include:

- \*
- windows
- osx
- linux

An architecture suffix (`-x32`, `-x64`) is supported
but no longer relevant in today's world.

Note that this is an optional property:
if your releases supports all platforms you do not specify `"*"`,
but instead omit the property completely.

```json
{
	"details": "https://github.com/wbond/sublime_alignment",
	"releases": [
		{
			"platforms": ["osx", "linux"],
			"sublime_text": "*",
			"tags": "posix-"
		},
		{
			"platforms": "windows",
			"sublime_text": "*",
			"tags": "win32-"
		}
	]
}
```

### `base`

If for some reason one of the releases is from a different repository
than the top-level "details" key,
a "base" key may be specified in the release.
This repository will then be used for the matching tags.

```json
{
	"details": "https://github.com/wbond/sublime_alignment",
	"releases": [
		{
			"base": "https://github.com/wbond/sublime_alignment",
			"sublime_text": "<3000",
			"tags": true
		},
		{
			"sublime_text": ">=3000",
			"tags": true
		}
	]
}
```

### `version`, `url`, `date`

> For manually created release zips files,
> each individual "version" needs to be specified with
> a "url" to the zip, and a "date".
> Because this requires an update of the repository file for each release
> it is not allowed in the default repository.

Reviewer: ```suggestion
For manually created release zip files,
each individual version" needs to be specified with
a "url" to the zip, and a "date".
Because this requires an update of the repository file for each release,
it is not allowed in the default repository.
```

Versions needs to be numbered in accordance with [semver](https://semver.org). 

The URL needs to be a zip file containing the package.
It is permissible for the zip file to contain a single root folder
with any name.
All files will be extracted out of this single root folder.

> This allows zip files from GitHub, GitLab and Bitbucket to be used a sources.

Reviewer: ```suggestion
This allows zip files from GitHub, GitLab, and Bitbucket to be used as sources.
```

The URL can be a relative path

> from the location of the repository.json file that specifies it.

Reviewer: ```suggestion
from the location of the `repository.json` file that specifies it.
```

The date must be in the form "YYYY-MM-DD HH:MM:SS" and should be UTC.

```json
{
	"packages": [
		{
			"details": "https://github.com/wbond/sublime_alignment",
			"releases": [
				{
					"version": "2.0.0",
					"url": "https://codeload.github.com/wbond/sublime_alignment/zip/v2.0.0",
					"date": "2011-09-18 20:12:41",
					"sublime_text": "*"
				}
			]
		},
		{
			"details": "../alignment",
			"releases": [
				{
					"version": "2.0.0",
					"url": "./downloads/alignment-2.0.0.zip",
					"date": "2011-09-18 20:12:41",
					"sublime_text": "*"
				}
			]
		}
	]
}
```

### `asset`

If your package requires a build or compile step,
you might want to provide your releases via pre-compiled zip archives
or sublime-package files.

> To do so, specify an `asset` key describing

Reviewer: Some of the JSON keys in this page use code spans, where others are double-quoted. Pick one or the other.

Author: I think I have this covered now 🤞🏻

		},
		{
			"platforms": ["linux-x64"],
			"python_versions": ["3.3"],
			"version": "4.3.4",
			"url": "https://files.pythonhosted.org/packages/8a/0f/5221822805edf3fc13e85c278de6451a5c08d0fd67e2c86e67e48b683a20/coverage-4.3.4-cp33-cp33m-manylinux1_x86_64.whl"
		},
		{
			"platforms": ["osx-x64"],
			"python_versions": ["3.3"],
			"version": "4.3.4",
			"url": "https://files.pythonhosted.org/packages/ac/dc/3e2d996c440a1a589f3323e806cf96d3c64650579483c3798ef2ea34b51a/coverage-4.3.4-cp33-cp33m-macosx_10_10_x86_64.whl"
		},
		{
			"platforms": ["windows-x32"],
			"python_versions": ["3.3"],
			"version": "4.2.0",
			"url": "https://files.pythonhosted.org/packages/a0/34/1185348cc5c541bbdf107438f0f0ea9df5d9a4233a974e9228b6ee815489/coverage-4.2-cp33-cp33m-win32.whl"
		},
		{
			"platforms": ["windows-x64"],
			"python_versions": ["3.3"],
			"version": "4.2.0",
			"url": "https://files.pythonhosted.org/packages/b1/55/02815cb8abb091033abb979ebde5122bb33b85c5987dede9ccd019033d19/coverage-4.2-cp33-cp33m-win_amd64.whl"
		}
	]
}
```

> Latest wheel files can be fetched directly from PyPI,
> using their official package URL,
> even in a mix with releases in legacy dependency format.

Reviewer: ```suggestion
The latest wheel files can be fetched directly from PyPI,
using their official package URLs,
even in a mix with releases in the legacy dependency format.
```

> ```json

Reviewer: ```suggestion
```jsonc
```

(If we support that)

{
	"name": "coverage",
	"description": "coverage.py - http://coverage.readthedocs.org/en/latest/",
	"author": "wbond",
	"issues": "https://github.com/codexns/sublime-coverage/issues",
	"releases": [
		{
			"base": "https://github.com/codexns/sublime-coverage",
			"platforms": ["*"],
			"python_versions": ["3.3"],
			"tags": true
		},
		{
			"base": "https://pypi.org/project/coverage",
			"asset": "coverage-*-cp38-cp38-win_amd64.whl",
			"platforms": ["windows-x64"],
			"python_versions": ["3.8"]
		},
		{
			"base": "https://pypi.org/project/coverage",
			"asset": "coverage-*-cp38-cp38-manylinux_2_17_aarch64*.whl",
			"platforms": ["linux-arm64"],
			"python_versions": ["3.8"]
		},
		{
			"base": "https://pypi.org/project/coverage",
			"asset": "coverage-*-cp38-cp38-manylinux_2_5_x86_64*.whl",
			"platforms": ["linux-x64"],
			"python_versions": ["3.8"]
		},
		{
			"base": "https://pypi.org/project/coverage",
			"asset": "coverage-*-cp38-cp38-macosx_11_0_arm64.whl",
			"platforms": ["osx-arm64"],
			"python_versions": ["3.8"]
		},

> 		{	// fixed versions is specified by URL.

Reviewer: ```suggestion
		{	// A fixed version is specified by URL.
```
or
```suggestion
		{	// Fixed versions are specified by URL.
```

			"base": "https://pypi.org/project/coverage/7.0.3",
			"asset": "coverage-*-cp38-cp38-macosx_10_9_x86_64.whl",
			"platforms": ["osx-x64"],
			"python_versions": ["3.8"]
		}
	]
}
```

Legacy dependencies can use download assets
(see the previous topic around assets)
to ship platform specific archives.
Download size for dependencies can be reduced this way,
as only compatible binaries need to be downloaded.  
Note: Legacy dependency format is fully supported
also for ST4 and Python 3.8,
but it is recommended to ship dependencies as wheels.

> ```json

Reviewer: ```suggestion
```jsonc
```
(Again, if we support it)
